Tools/gdb/libpython.py

```diff
@@ -890,11 +892,18 @@ def iter_locals(self):
             pyop_name = PyObjectPtr.from_pyobject_ptr(self.co_localsplusnames[i])
             yield (pyop_name, pyop_value)
 
+    def _f_specials(self, index):
+        f_valuestack = self.field('f_valuestack')
+        return PyObjectPtr.from_pyobject_ptr(f_valuestack[index - FRAME_SPECIALS_SIZE])
+
+
     def _f_globals(self):
-        f_localsplus = self.field('f_localsptr')
-        nlocalsplus = int_from_int(self.co.field('co_nlocalsplus'))
-        index = nlocalsplus + FRAME_SPECIALS_GLOBAL_OFFSET
-        return PyObjectPtr.from_pyobject_ptr(f_localsplus[index])
+        return self._f_specials(FRAME_SPECIALS_GLOBAL_OFFSET)
+
+    def _f_code(self):
+        f_valuestack = self.field('f_valuestack')
+        code = f_valuestack[FRAME_SPECIALS_CODE_OFFSET - FRAME_SPECIALS_SIZE]
+        return PyCodeObjectPtr.from_pyobject_ptr(code)
```

Comment:
I suppose it doesn't matter much but it's too bad you can't use `self._f_specials()` here.  With a small adjustment to `_f_specials()` this code becomes simpler and more uniform.

For example, let the caller pass the wrapper to use (and default to `PyObjectPtr`):

```suggestion
        return self._f_specials(FRAME_SPECIALS_CODE_OFFSET, PyCodeObjectPtr)
```

...or let `_f_specials()` decide based on the index (since this code is already so specific and tightly coupled):

```suggestion
        return self._f_specials(FRAME_SPECIALS_CODE_OFFSET)
```